core/src/com/unciv/logic/city/CityInfo.kt

```diff
@@ -175,60 +180,85 @@ class CityInfo {
+                (prefix.repeat(number) + cityName) !in usedCityNames
+            }
+            if (candidate != null) return prefix.repeat(number) + candidate
+            number++
         }
-
-        name = cityNamePrefix + cityName
     }
 
-    private fun borrowCityName(): String {
-        val usedCityNames =
-            civInfo.gameInfo.civilizations.flatMap { it.cities.map { city -> city.name } }
-        // We take the last unused city name for each other civ in this game, skipping civs whose
-        // names are exhausted, and choose a random one from that pool if it's not empty.
-        var newNames = civInfo.gameInfo.civilizations
-            .filter { it.isMajorCiv() && it != civInfo }
-            .mapNotNull {
-                it.nation.cities
-                    .lastOrNull { city -> city !in usedCityNames }
-            }
-        if (newNames.isNotEmpty()) {
-            return newNames.random()
+    /**
+     * Generates and returns a city name borrowed from another nation.
+     *
+     * @param usedCityNames Every city name in use including foreign cities.
+     * @return A borrowed city name in [String].
+     */
+    private fun generateBorrowedCityName(usedCityNames: List<String>): String {
+        // Attempt to return a random name from the pool of last unused city name for every *other*
+        // nation in this game, skipping nations whose names are exhausted
+        val otherNations = civInfo.gameInfo.civilizations.filter { nation ->
+            nation.isMajorCiv() && nation != civInfo
+        }
+        var unusedCityNames = otherNations.mapNotNull {
+            it.nation.cities.lastOrNull { name -> name !in usedCityNames }
         }
+        if (unusedCityNames.isNotEmpty()) return unusedCityNames.random()
 
         // As per fandom wiki, once the names from the other nations in the game are exhausted,
-        // names are taken from the rest of the nations in the ruleset
-        newNames = getRuleset()
-            .nations
-            .filter { it.key !in civInfo.gameInfo.civilizations.map { civ -> civ.nation.name } }
-            .values
-            .map {
-                it.cities
-                    .filter { city -> city !in usedCityNames }
-            }.flatten()
-        if (newNames.isNotEmpty()) {
-            return newNames.random()
-        }
-        // If for some reason we have used every single city name in the game,
-        // (are we using some sort of baserule mod without city names?)
+        // names are taken from the rest of the nations in the rule set
+        val missingNations = getRuleset().nations.filter { nation ->
+            nation.key !in civInfo.gameInfo.civilizations.map { civ -> civ.nation.name }
+        }
+        unusedCityNames = missingNations.values.flatMap { nation ->
+            nation.cities.filter { city -> city !in usedCityNames }
+        }
+        if (unusedCityNames.isNotEmpty()) return unusedCityNames.random()
```

Comment:
I personally do not necessarily feel that rewriting this function by adding additional variables but keeping the logic the same improves readability. This is partly as I've written this function myself. If anyone else does feel that way, I'm fine with changing it, but as it is I do not necessarily see the added value of adding additional vars and vals that are only ever used once :)